# Import config file & parameters

configfile: 'config_charlotte.yaml'

# Import paths from config file

DATAPATH=config['datapath']


rule all:
    input:
        # expand("fastqc/{sample}_fastqc.zip", sample=config["samples"]),
        # expand("fastqc/{sample}_fastqc.html", sample=config["samples"]),
        # expand("trim_galore/{sample}_trimmed.fq.gz", sample=config["samples"]),
        # expand("mapped/{sample}.", sample=config["samples"]),
        expand("mapped/{sample}.Aligned.sortedByCoord.out.bam.bai", sample=config["samples"])

rule fastqc:
    input:
        fasta=expand(DATAPATH + "{sample}.fq.gz", sample=config["samples"])
    output:
        "fastqc/{sample}_fastqc.zip",
        "fastqc/{sample}_fastqc.html"
    threads:
        4
    shell:
        "fastqc -t {threads} {input.fasta} --outdir fastqc"


rule trimGalore:
    input:
        fasta=expand(DATAPATH + "{sample}.fq.gz", sample=config["samples"])
    output:
        "trim_galore/{sample}_trimmed.fq.gz",
    params:
        basic = "-q 20 --gzip --length 16 -o trim_galore --fastqc",
        adaptor = "" if config["trim_galore_params"] else "-a "+config["trim_galore_params"]
    shell:
        "trim_galore {params.basic} {params.adaptor} {input.fasta}"

# ——————————————————
# Nobby begin

rule remove_rRNA_tRNA:
    input:
        expand("trim_galore/{sample}_trimmed.fq.gz", sample=config["samples"])
    output:
        expand("rRNA_tRNA_removed/{sample}.rRNA_tRNA_removed.fq.gz", sample=config["samples"])
    log:
        expand("logs/{sample}.rRNA_tRNA_removal.log", sample=config["samples"])
    params:
        bt2_index=config['bt2_rRNA_tRNA_index'],
        threads=8
    shell:
        "bowtie2 -p {params.threads} --un-gz {output} -x {params.bt2_index} -U {input} > /dev/null 2> {log}"

rule map_star:
    input:
        expand("rRNA_tRNA_removed/{sample}.rRNA_tRNA_removed.fq.gz", sample=config["samples"])
    output:
        bam = expand("mapped/{sample}.Aligned.sortedByCoord.out.bam", sample=config["samples"])
    params:
        star_index=config['star_index'],
        prefix = expand("mapped/{sample}.", sample=config["samples"]),
        threads=8
    shell:
        "STAR --genomeDir {params.star_index} --readFilesIn {input} --readFilesCommand zcat --outFileNamePrefix {params.prefix} --runThreadN {params.threads} --genomeLoad NoSharedMemory --outFilterMultimapNmax 1 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1  --outFilterMismatchNmax 999 --outFilterMismatchNoverReadLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --outFilterType BySJout --outSAMattributes All --outSAMstrandField intronMotif --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --sjdbScore 1 --limitBAMsortRAM 60000000000"

rule sambamba_index:
    input:
        expand("mapped/{sample}.Aligned.sortedByCoord.out.bam", sample=config["samples"])
    output:
        expand("mapped/{sample}.Aligned.sortedByCoord.out.bam.bai", sample=config["samples"])
    params:
        threads=8
    shell:
        "sambamba index -t {params.threads} {input} {output}"
